command/agent/auth/auth.go

```diff
@@ -235,6 +241,7 @@ func (ah *AuthHandler) Run(ctx context.Context, am AuthMethod) error {
 		// This should only happen if there's no preloaded token (regular auto-auth login)
 		//  or if a preloaded token has expired and is now switching to auto-auth.
 		if secret.Auth == nil {
+			clientToUse = clientToUse.WithResponseCallbacks(api.RecordState(&ah.vaultIndex))
```

Comment:
It might be safer, instead of assigning to clientToUse, to have the clone with the response callback only get used for this one write.  Otherwise any subsequent use of clientToUse could update vaultIndex again, possibly racing with our later read of that field.

Reply:
This makes sense to me. I did try an implementation with a lock and while I removed it afterwards, as it wasn't strictly necessary, this approach protects us from the code evolving into races later. Thanks for the suggestion!